### PR TITLE
Corrected Currency Display: Enforce Two Decimal Places in Amounts

### PR DESCRIPTION
**Pull request by abzokhattab**
### Fixed Issues

$ https://github.com/Expensify/App/issues/34894
$ https://github.com/Expensify/App/issues/34894#issuecomment-1904253989

### Tests

- [X] Verify that no errors appear in the JS console

1. Navigate to FAB (Floating Action Button) 
2. Select Request Money
3. Select manual
4. Enter an amount that has a last decimal digit as 0 (e.g. 23.20)
5. click on Next > go back using the header back button
6. Ensure the amount shown to the user is the same as in step 4

### Offline tests

1. Navigate to FAB (Floating Action Button) 
2. Select Request Money
3. Select manual
4. Enter an amount that has a last decimal digit as 0 (e.g. 23.20)
5. click on Next > go back using the header back button
6. Ensure the amount shown to the user is the same as in step 4

### QA Steps

- [x] Verify that no errors appear in the JS console

1. Navigate to FAB (Floating Action Button) 
2. Select Request Money
3. Select manual
4. Enter an amount that has a last decimal digit as 0 (e.g. 23.20)
5. click on Next > go back using the header back button
6. Ensure the amount shown to the user is the same as in step 4

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
        - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/59809993/594266c0-0e05-4a4c-9308-e2d4566f7d17

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/59809993/dffddab9-b069-4636-9906-b7083379a31d

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/59809993/594266c0-0e05-4a4c-9308-e2d4566f7d17

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/59809993/dffddab9-b069-4636-9906-b7083379a31d

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/59809993/c08481db-accc-47c3-8e34-03f6e40fbd56

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/59809993/1e7e6e9f-f599-4561-b183-082b3f7d6369

</details>

### CODE REVIEW

**Review by eVoloshchak — COMMENTED** [review]
Bug:

1. Open any 1:1 chat
2. Press '+' button -> Request Money
3. Enter an amount without the decimal point (i.e. "5")
4. Click "Next"
5. Click "Back"
6. Notice the amount is now "5.00"


https://github.com/Expensify/App/assets/9059945/68fb2648-0991-43ff-b78d-da4a91fbbbad

This is kind of an edge case, [this](https://github.com/Expensify/App/issues/34894#issuecomment-1915238369) comment explains it well


> you're not wrong https://github.com/Expensify/App/issues/34894#issuecomment-1914468621, it's not quite consistent, but I think if the user has JUST entered a value, we'd want to maintain it as is, making it easier for the user to fix if needed. This is also how Square Cash and Venmo do it, as a reference.

**Comment by eVoloshchak** [comment]
This is still present
It should work this way for any number, not just for numbers without the decimal point
https://github.com/Expensify/App/assets/9059945/9e1df7c0-d730-4ed8-877f-4d975db2cc06

**Comment by eVoloshchak** [comment]
> why do you consider the case mentioned in the video attached as a bug, i think it makes sense to be converted to 8.90

@abzokhattab, this isn't a bug per se, but I agree with [this](https://github.com/Expensify/App/issues/34894#issuecomment-1915238369) comment 
> but I think if the user has JUST entered a value, we'd want to maintain it as is, making it easier for the user to fix if needed. This is also how Square Cash and Venmo do it, as a reference.

The key here is we shouldn't change the amount if user JUST entered it. Say, you enter 8.8, click 'Next' and immediatelly come back wanting to change it. If it's immediately converted to 8.80, you have one additional keypress to do (and it's much worse for 8.00)

**Comment by eVoloshchak** [comment]
Video below demonstrates the desired behavior
https://github.com/Expensify/App/assets/9059945/460be79a-3b96-4eb5-8b25-dc7f8696e889

Essentially, if you enter an amount, press next and then immediately press 'Back', amount should stay the same. 
But if you enter an amount, press next and then press on amount to edit it, it should be displayed with two decimal places

**Comment by eVoloshchak** [comment]
@abzokhattab, this doesn't seem to work correctly, I can still see the same bug

https://github.com/Expensify/App/assets/9059945/c866ac70-0fef-4a17-8f9a-6e27322180c6

**Comment by abzokhattab** [comment]
kindly reminder @eVoloshchak

**Comment by dangrous** [comment]
@eVoloshchak will you be able to look at this today or tomorrow? Thank you!

**Comment by eVoloshchak** [comment]
Still haven't found the approach that would work correctly for both cases, will go over this tomorrow

**Comment by abzokhattab** [comment]
Thanks will let you know if I was able to find a solution as well.

**Comment by eVoloshchak** [comment]
Ok, I think I found a solution
What we need to achieve: 
1. format the amount with two decimal places when you're on edit amount screen
2. display the amount unchanged (without two decimal places) if navigated back to the first screen of the request money flow

We can utilize `useRoute` to get the current route
const currentRoute = useRoute();

And then based on whether it's [MONEY_REQUEST.STEP_AMOUNT](https://github.com/Expensify/App/blob/63dee4e1b31f8801ba9c8ea997b8dd4d54712aa1/src/SCREENS.ts#L141) or one of [TAB_REQUEST](https://github.com/Expensify/App/blob/63dee4e1b31f8801ba9c8ea997b8dd4d54712aa1/src/CONST.ts#L3081), we can pass a boolean to `convertToFrontendAmount` and format the amount accordingly

@abzokhattab, let me know what you think and how this works for other occurrences of `MoneyRequestAmountForm`
image

**Comment by abzokhattab** [comment]
Good job i added a check that changes the amount to two decimals only if the form is edit amount form,

Here is the test video:


https://github.com/Expensify/App/assets/59809993/81ff5580-9626-4f10-8937-e57c59fe692b

**Review by eVoloshchak — COMMENTED** [review]
Bug:

1. Navigate to FAB (Floating Action Button)
2. Select Request Money
3. Select manual
4. Enter an amount, click next
5. Click on Amount menu item to edit the amount
6. Enter the updated amount and click Save
7. Navigate back using the header back button
8. Notice the amount isn't updated


https://github.com/Expensify/App/assets/9059945/59ab1486-d06b-427b-b66c-8ddf2bae0ce7

File: src/libs/Permissions.ts [hunk 1]
```diff
@@ -3,6 +3,7 @@ import CONST from '@src/CONST';
 import type Beta from '@src/types/onyx/Beta';
 
 function canUseAllBetas(betas: OnyxEntry<Beta[]>): boolean {
+    return true;
```
Comment (eVoloshchak) [inline src/libs/Permissions.ts:6]: This was added here accidentally it seems

**Review by eVoloshchak — APPROVED** [review]
Looks good and tests well
@abzokhattab, thank you for the awesome work and patience on this one!

**Comment by dangrous** [comment]
Okay so this is looking good but one update - I think we had discussed (and maybe it wasn't clear, so if so I apologize) [keeping the value the same as how the user entered it](https://github.com/Expensify/App/issues/34894#issuecomment-1915238369), if they just entered it and it is a whole currency value (so just `$23`, vs. `$23.20` or whatever). Can we work that into this solution? Or am I missing something on the PR - which I totally could be.

Thank you!

**Comment by dangrous** [comment]
Okay, yeah - I think it's a small price to pay for the consistency. Thank you for the explanation! Let's keep it as you have it. Will give this a final look through tomorrow.

**Review by dangrous — CHANGES_REQUESTED** [review]
This looks good! Can we also add some CurrencyUtilTests for the new function - convertToFrontendAmountAsString, including both numbers and null values to get the empty string? Then I think we should be ready to go after a bit of final testing I need to do!

**Comment by abzokhattab** [comment]
@dangrous  okay i added some tests for the edge cases ... please let me now if you have any other comments.

File: tests/unit/CurrencyUtilsTest.ts [hunk 2]
```diff
@@ -123,8 +123,12 @@ describe('CurrencyUtils', () => {
             [2550, '25.50'],
             [25, '0.25'],
             [2500.5, '25.00'],
-        ])('Correctly converts %s to amount in units handled in frontend as a string', (amount, expectedResult) => {
-            expect(CurrencyUtils.convertToFrontendAmountAsString(amount)).toBe(expectedResult);
+            [null, ''],
+            [undefined, ''],
+            ['', '0.00'],
```
Comment (dangrous) [inline tests/unit/CurrencyUtilsTest.ts:128]: we don't accept an empty string in the function as of now

**Review by dangrous — CHANGES_REQUESTED** [review]
One minor change just to make typescript happy!

File: tests/unit/CurrencyUtilsTest.ts (hunk 2 above)
Comment (abzokhattab) [reply tests/unit/CurrencyUtilsTest.ts:128]: so should i just remove the empty string test?

Comment (abzokhattab) [reply tests/unit/CurrencyUtilsTest.ts]: i removed it in the recent commit .. let me know what you think

**Review by dangrous — COMMENTED** [review]
cool, this looks good to me!

**Review by dangrous — DISMISSED** [review]

**Comment by abzokhattab** [comment]
There we go @dangrous

**Comment by dangrous** [comment]
arggghhh sorry. One more time? I'll keep this tab open so it hopefully doesn't happen again

**Comment by abzokhattab** [comment]
> arggghhh sorry. One more time? I'll keep this tab open so it hopefully doesn't happen again

😂😂😂😂 hopefully this is the last time.

**Review by dangrous — APPROVED** [review]
Let's do it!

**Pull request merged**